### PR TITLE
Refactor - Core/Container/Media Control

### PR DESCRIPTION
**Pull request by joaopaulovieira**
## Summary 

Major refactor regarding Core/Containers/MediaControl relationships. The goal of the changes is to have a more sane MediaControl overriding and some other small changes in the components signatures. 

This was planned for us to move in the direction of having a more customizable MediaControl, now turned into a UICorePlugin itself, making it easier to be overriden as well. It also contributes in having a better separation between components in order to have a more minimalistic Clappr core (which we'll call clappr-core) and separated plugins to have the player built in a mix-and-match style, having only the plugins the user needs.

### Core

- Added `activeContainer` and `activePlayback` properties to replace `getCurrentContainer`/`getCurrentPlayback`. The event `CORE_ACTIVE_CONTAINER_CHANGED` was created, being triggered whenever the `activeContainer` setter is called. This new event replaces `MEDIACONTROL_CONTAINERCHANGED`. This is a *breaking change*.
- Removed `MediaControl` instantiation and initialization (it was converted into a `UICorePlugin`).
- Deprecated `getCurrentContainer`/`getCurrentPlayback` methods. The implementations were updated to use the newer `activeContainer`/`activePlayback`.
- Create `CORE_MOUSE_MOVE` and `CORE_MOUSE_LEAVE` events and trigger when DOM mouse events occurs on `core` element.
- Removes all unnecessary `mediacontrol` references, since it has been turned into a plugin.
- Added `mediaControl` getter to keep compatibility, using a dummy component if there's no MediaControl plugin configured.

### MediaControl

- Changed `MediaControl` into an `UICorePlugin`. Now in order to replace the MediaControl with another custom implementation, instead of passing an option `mediacontrol.external` we just need to pass the replacement MediaControl in the `plugins` list, keeping in mind that the plugin `name` property should be `media_control` to use the plugin overriding system correctly. This is a *breaking change*.
- Listen the`CORE_OPTIONS_CHANGE` event to set new options.
- Listen `CORE_MOUSE_MOVE`/`CORE_MOUSE_LEAVE` for trigger show/hide actions.
- Call the `toggleFulscreen` on `core` when the full screen is triggered.

### UIObject/UIContainerPlugin/UICorePlugin
- Renamed method `remove` to `destroy` on `UIObject`.
- All visual plugins, as well as `UIObject` have now a consistent interface, with `destroy` removing them from DOM and liberating listeners, resources, etc.

### Other

- All built-in plugins have been updated accordingly.

### CODE REVIEW

**Comment by towerz** [comment]
This change was planned to prepare for the next iteration in the player architecture. 

Actually, it has been discussed for a while between most of the core team members, and we decided to split it into some phases. This is the first phase of the changes required for our architecture review and the reduction of the player footprint, as well as the project being split into smaller ones.

If anyone else wants to review this, we'll leave the PR open until Oct 22. @clappr/core, if anyone there wants to review it, please do. Just don't merge it yet.

File: src/components/player.js [hunk 1]
```diff
@@ -499,9 +508,7 @@ export default class Player extends BaseObject {
    * @return {Player} itself
    */
   setVolume(volume) {
-    if (this.core && this.core.mediaControl)
-      this.core.mediaControl.setVolume(volume)
-
+    this.core && this.core.mediaControl && this.core.mediaControl.setVolume(volume)
```
[the remark this answers is not in the document]
Comment (leandromoreira) [reply src/components/player.js]: I agree with @jhonatangomes and here we're setting a direct dependency through `core` and `core.mediaControl`

[the remark this answers is not in the document]
Comment (joaopaulovieira) [reply src/components/player.js]: My fault, I forgot that. 😅
Talking with @vagnervjs  and @jhonatangomes , we decided to remove these methods(`setVolume`/`getVolume`) from `player.js` and expose on `media_control.js` the volume get/set operations via external interface, keeping the method assignature for reasons of backward compatibility.

File: src/components/core/core.js [hunk 2]
```diff
@@ -65,6 +64,47 @@ export default class Core extends UIObject {
...
+   * @deprecated
+   * This property currently exists for retrocompatibility reasons.
+   * If you want to access the media control instance, use the method getPlugin('media_control').
+   */
+  get mediaControl() {
+    return this.getPlugin('media_control') || this.dummyMediaControl
+  }
+
+  get dummyMediaControl() {
+    return new UICorePlugin(this)
```
[the remark this answers is not in the document]
Comment (joaopaulovieira) [reply src/components/core/core.js]: Of course! I made the change in this commit de2d52d517f3cf9501564aa5f3a18225172dd447

File: src/base/ui_object.js [hunk 3]
```diff
@@ -122,10 +122,10 @@ export default class UIObject extends BaseObject {
...
   /**
    * removes the ui component from DOM
-   * @method remove
+   * @method destroy
    * @return {UIObject} itself
    */
-  remove() {
+  destroy() {
     this.$el.remove()
     this.stopListening()
```
Comment (afonsocdaniel) [inline src/base/ui_object.js]: What do you think about moving ```this.stopListening()``` to the *BaseObject*  and calls here the method ```super.destroy()```? With this change, we can also remove the method ```destroy``` from the others components that extends from ***BaseObject***.

Comment (towerz) [reply src/base/ui_object.js]: @afonsocdaniel as we agreeed, let's do this as a separate change.

**Pull request merged**